nucliadb_telemetry/nucliadb_telemetry/logs.py

```diff
@@ -0,0 +1,192 @@
+# For commercial licensing, contact us at info@nuclia.com.
+#
+# AGPL:
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+# inspired from datadog integrations like https://pypi.org/project/JSON-log-formatter/
+import logging
+from copy import copy
+from datetime import datetime
+from typing import Any
+
+import orjson
+import pydantic
+from opentelemetry import trace
+from opentelemetry.trace.span import INVALID_SPAN
+
+from nucliadb_telemetry.settings import LogSettings
+
+_BUILTIN_ATTRS = (
+    # list of all possible args
+    "args",
+    "asctime",
+    "created",
+    "exc_info",
+    "exc_text",
+    "filename",
+    "funcName",
+    "levelname",
+    "levelno",
+    "lineno",
+    "module",
+    "msecs",
+    "message",
+    "msg",
+    "name",
+    "pathname",
+    "process",
+    "processName",
+    "relativeCreated",
+    "stack_info",
+    "thread",
+    "threadName",
+)
+
+_OUTPUT_PROPERTIES = {
+    "filename": "filename",
+    "funcName": "funcName",
+    "levelname": "severity",
+    "lineno": "lineno",
+    "module": "module",
+    "name": "name",
+    "pathname": "pathname",
+}
+
+
+class JSONFormatter(logging.Formatter):
+    """
+    Formatter base json-log-formatter with pydantic support
+    """
+
+    def format(self, record: logging.LogRecord) -> str:
+        extra: dict[str, Any]
+        if isinstance(record.msg, dict):
+            extra = record.msg
+        elif isinstance(record.msg, pydantic.BaseModel):
+            extra = record.msg.dict()
+        else:
+            extra = {"message": record.getMessage()}
+        extra.update(self.extra_from_record(record))
+
+        self.fill_log_data(extra, record)
+
+        return orjson.dumps(extra, default=repr).decode("utf-8", errors="ignore")
+
+    def fill_log_data(self, data: dict[str, Any], record: logging.LogRecord) -> None:
+        if "time" not in data:
+            data["time"] = datetime.utcnow()
+
+        if record.exc_info:
+            data["exc_info"] = self.formatException(record.exc_info)
+
+        for property, output_property in _OUTPUT_PROPERTIES.items():
+            data[output_property] = getattr(record, property, None)
+
+        data["logging.googleapis.com/sourceLocation"] = {
+            "file": record.pathname,
+            "line": record.lineno,
+            "function": record.funcName,
+        }
```

Comment:
Why do we need this?
Isn't it already covered by `filename`, `lineno` and `funcName`?

Reply:
alright, tweaked, lmkwyt